### PR TITLE
Main readme redesign

### PR DESCRIPTION
**Pull request by YoelShoshan**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,45 +5,164 @@
 
 <img src="fuse/doc/FuseMedML-logo.png" alt="drawing" width="30%"/>
 
-# What is FuseMedML?
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply README.md:8]: And about that description, maybe boosting or enabling instead of encouraging.
Maybe we should also mention that it improves collaboration 
Just thoughts that I do not completely agree with :-) so feel free to ignore,

**Review by mosheraboh — CHANGES_REQUESTED** [review]
Great!
Few comments inline.
Maybe it is a good point to ask Micha (who is not that familiar with FuseMedML) to review?

File: README.md [hunk 2]
```diff
@@ -5,45 +5,164 @@
...
+An example metric that can be used
+```python
+MetricAUCROC(
+    pred='model.output', # input - model prediction scores
+    target='data.label'  # input - ground truth labels
+)
+```
 
-* **Compatibility with alternative frameworks**
+Note that several components return answers directly and not write it into the nested dictionary. This is perfectly fine, and to allow maximum flexibility we do not require any usage of output path keys.
```
Comment (mmdanziger) [inline README.md:77]: Do we need this? In any event, readers don't have the understanding to parse this statement yet. Return? We haven't seen any returns or outputs yet at all. Maybe this can come after the nested dict if the nested dict is pushed lower in the sequence.
Even better, just include this option as part of a live example, in the spirit of "show don't tell".

File: README.md [hunk 3]
```diff
@@ -5,45 +5,164 @@
...
+* loss
+* multihead model
+
+### Creating a custom FuseMedML component
+
+Creating custom FuseMedML components is easy - in the following example we add a new data processing operator:
+
+
+A data pipeline operator
+```python
```
Comment (mmdanziger) [inline README.md:89]: I like this! Well motivated and explicit. Only problem is I don't know how it fits in with the other pieces. I assume there's some kind of pipeline object that this can be added as a list in. Can that be shown also?

File: README.md [hunk 4]
```diff
@@ -5,45 +5,164 @@
...
+Arguments can be freely named, and you don't even have to write anything to the nested dict.
+Some FuseMedML components return a value directly - for example, loss functions.
+
+## 2. "Batteries included" key components, built using the same design concept
+
+### **fuse.data** - A **declarative** super flexible data processing pipeline
+* Easy dealing with complex multi modality scenario
+* Advanced caching, including periodic audits to automatically detect stale caches
+* Default ready-to-use Dataset and Sampler classes
+* See detailed introduction [here](https://github.com/IBM/fuse-med-ml/tree/master/fuse/data/README.md)
```
Comment (mmdanziger) [inline README.md:122]: Is there any way to link to examples for these items?

File: README.md [hunk 5]
```diff
@@ -5,45 +5,164 @@
...
+Some components depend on pytorch. For example, ```fuse.data``` is oriented towards pytorch DataSet, DataLoader, DataSampler etc.
+```fuse.dl``` makes heavy usage of pytorch models.
+Some components do not depend on any specific DL library - for example ```fuse.eval```.
+
+Broadly speaking, the supported DL libraries are:
+* "Pure" [pytorch](https://pytorch.org/)
+* [pytorch-lightning](https://www.pytorchlightning.ai/)
+
+Before you ask - **pytorch-lightning and FuseMedML play along very nicely and have in practice orthogonal and additive benefits :)**  
+See [Simple FuseMedML + PytorchLightning Example](./examples/fuse_examples/imaging/classification/mnist/run_mnist.py) for simple supervised learning cases, and [this example ](./examples/fuse_examples/imaging/classification/mnist/run_mnist_custom_pl_imp.py) for completely custom usage of pytorch-lightning and FuseMedML - useful for advanced scenarios such as Reinforcement Learning and generative models.
```
Comment (mmdanziger) [inline README.md:138]: In my mind this is **the most important** point and it deserves more discussion. Specifically, given that there are already a lot of DL libraries, the dev needs to be convinced to pick fuse. For example lightning addresses this question vis-a-vis vanilla torch in two ways: 
1. Via the super cool gif showing how you can transform complex vanilla torch code into simple lightning code.
2. An explicit section https://github.com/Lightning-AI/lightning#advantages-over-unstructured-pytorch

That is what is missing here. Ok, I **could** use fuse with lightning. But why add fuse if I already have fluency with lightning?

Also, lightning is not the only kid in town. There's https://github.com/project-monai/monai and probably many more that I am not familiar with. Is fuse better than all of them for every ML projects? Or are some projects better suited to vanilla torch or lightning? How should I decide? What exactly is it that fuse can do better than the others, given that everyone covers architecture, loss, training and evaluation in some way or other? Can you show me?

These are the essential questions, in my opinion.

File: README.md [hunk 6]
```diff
@@ -5,45 +5,164 @@
...
+
+## Domain Extensions
+fuse-med-ml, the core library, is completely domain agnostic!
+Domain extensions are optionally installable packages that deal with specific (sub) domains. For example:
+
+* **fuseimg** which was battletested in many medical imaging related projects (different organs, imaging modalities, tasks, etc.)
+* **fusedrug** which focuses on molecular biology and chemistry - prediction, generation and more
```
Comment (mmdanziger) [inline README.md:145]: links?

File: README.md [hunk 7]
```diff
@@ -5,45 +5,164 @@
...
+
+Domain extensions contain concrete implementation of components and components parts within the relevant domain, for example:  
+* Data pipeline operations - for example, a 3d affine transformation of a 3d image
+* Evaluation metrics - for example, a custom metric evaluating docking of a potential drug with a protein target
+* Loss functions - for example, a custom segmentation evaluation loss
+
```
Comment (mmdanziger) [inline README.md:151]: links?

File: README.md [hunk 8]
```diff
@@ -5,45 +5,164 @@
...
+    data #everything related to datasets, samplers, data processing pipeline Ops, etc.
+    dl #everything related to deep learning architectures, optimizers, loss functions etc.
+    eval #evaluation metrics
+    utils #any utilities
+```
+
+You are highly encouraged to create additional domain extensions and/or contribute to the existing ones!
+There's no need to wait for any approval, you can create domain extensions on your own repos right away
+
+Note - in general, we find it helpful to follow the same directory structure shown above even in small and specific research projects that use FuseMedML for consistency and easy landing for newcomers into your project :)
```
Comment (mmdanziger) [inline README.md:164]: Do we have examples of such projects that depend on fuse but are managed independently?

File: README.md [hunk 9]
```diff
@@ -5,45 +5,164 @@
...
+See a short introduction video (3 minutes) to how FuseMedML components work:  
 
-FuseMedML is an outstanding collaboration framework that allows you to rerun an experiment or reuse some of the capabilities originally written for different projects—all with minimal effort.\
-Using FuseMedML, you can write generic components that can be easily shared between projects in a plug & play manner, simplifying sharing and collaboration.
+https://user-images.githubusercontent.com/7043815/177197158-d3ea0736-629e-4dcb-bd5e-666993fbcfa2.mp4
 
-The framework’s unique software design provides many advantages, making it an ideal framework for deep-learning research and development in medical imaging:
 
-* **Rapid development** -
+### Examples - using FuseMedML-style components
```
Comment (mmdanziger) [inline README.md:43]: Ok, this is it, your chance to show the readers why fuse is the awesomest thing ever. But you need to be careful. Showing new syntactical requirements without showing their payoff is not convincing. 
Learning to encode things with a new `NDict` format, wrapping losses and metrics in fuse classes--these are not features, they are costs for the dev. The feature is what they enable. 

When do we get to see that? Can we see it sooner 😃 ?

**Review by mmdanziger — COMMENTED** [review]
Hello, Efrat asked me to provide comments so I've tried my best here. I apologize in advance if I've misunderstood things or if the comments are not useful.

In general, I think you did a good job at pivoting in the direction of devs and dev interests. That's who you need to convince so it makes sense to speak their language.

If I could highlight two main points that I think could be better it would be:

1. Show don't tell. This is standard advice for fiction authors but works here also. Runnable code samples are the ultimate "show" on github.
2. Differentiate from other libraries. After spending hours reading through the readme I still don't understand what it is that fuse does better than lightning, monai etc. I added detailed points also.
3. Remember, your reader knows nothing about fuse except what you told him in the preceding lines of this document.

After reading through the document in detail, and knowing what I know from previous interactions, I have a better sense of the value that fuse offers, and it is impressive.

File: README.md [hunk 10]
```diff
@@ -5,45 +5,164 @@
 
 <img src="fuse/doc/FuseMedML-logo.png" alt="drawing" width="30%"/>
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply README.md:6]: correct

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply README.md:8]: done

File: README.md [hunk 11]
```diff
@@ -5,45 +5,164 @@
...
 
-# What is FuseMedML?
-FuseMedML is an open-source python-based framework designed to enhance collaboration and accelerate discoveries in **F**used **M**edical data through advanced **M**achine **L**earning technologies. Initial version is PyTorch-based and focuses on deep learning on medical imaging.
+# Effective Code Reuse across ML projects!
+**Battle tested in the medical domain:** Medical Imaging, Computational Molecular Biochem and EHR Clinical Data Analysis
+
+Skip to [install instructions section](#Installation)
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply README.md:11]: added

File: README.md [hunk 12]
```diff
@@ -5,45 +5,164 @@
...
-@misc{https://doi.org/10.5281/zenodo.5146491,
-  doi = {10.5281/ZENODO.5146491},
-  url = {https://zenodo.org/record/5146491},
-  author = {IBM Research,  Haifa},
-  title = {FuseMedML: https://github.com/IBM/fuse-med-ml},
-  publisher = {Zenodo},
-  year = {2021},
-  copyright = {Apache License 2.0}
-}
+* metrics
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply README.md:81]: done

File: README.md [hunk 13]
```diff
@@ -63,8 +182,7 @@ $ pip install fuse-med-ml
  FuseMedML supports Python 3.7 or later and PyTorch 1.5 or later. A full list of dependencies can be found in [**requirements.txt**](https://github.com/IBM/fuse-med-ml/tree/master/requirements.txt).
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply README.md]: done

File: README.md [hunk 14]
```diff
@@ -63,8 +182,7 @@ $ pip install fuse-med-ml
...
  
 
-# Ready to get started?
-## FuseMedML from the ground up
+# FuseMedML from the ground up
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply README.md:197]: done

File: README.md [hunk 15]
```diff
@@ -84,11 +202,8 @@ $ pip install fuse-med-ml
...
 
-
 ## Evaluation package
-The evaluation package of FuseMedML (fuse.eval) is a standalone library for evaluating ML models which not necessarily trained with FuseMedML.  
-The package includes collection of off-the-shelf metrics and utilities such as calibration, thresholding, model comparison and more.
-Details and examples can be found [here](https://github.com/IBM/fuse-med-ml/tree/master/fuse/eval/README.md)   
+ 
 
 
 ## Community support
```
[the remark this answers is not in the document]
Comment (YoelShoshan) [reply README.md:212]: done

**Review by mosheraboh — APPROVED** [review]
Approving, let's continue to address the comments after the release.

**Review by mosheraboh — DISMISSED** [review]

**Review by mosheraboh — APPROVED** [review]
